c/extras/py_ftrl.cc

```diff
@@ -24,60 +24,60 @@
 #include "python/int.h"
 #include "python/tuple.h"
 #include "extras/py_ftrl.h"
-
 namespace py {
 
+
 PKArgs Ftrl::Type::args___init__(1, 0, 7, false, false,
                                  {"params", "alpha", "beta", "lambda1",
                                  "lambda2", "d", "n_epochs", "inter"},
                                  "__init__", nullptr);
 
-std::vector<strpair> Ftrl::params_fields_info = {
-  strpair("alpha", "`alpha` in per-coordinate FTRL-Proximal algorithm"),
-  strpair("beta", "`beta` in per-coordinate FTRL-Proximal algorithm"),
-  strpair("lambda1", "L1 regularization parameter"),
-  strpair("lambda2", "L1 regularization parameter"),
-  strpair("d", "Number of bins to be used for the hashing trick"),
-  strpair("n_epochs", "Number of epochs to train a model for"),
-  strpair("inter", "Parameter that controls if feature interactions to be used "
-                   "or not")
-};
-strpair Ftrl::params_info = strpair("Params", "FTRL model parameters");
-onamedtupletype Ftrl::params_nttype(Ftrl::params_info,
-                                    Ftrl::params_fields_info);
+static const char* doc_alpha    = "`alpha` in per-coordinate FTRL-Proximal algorithm";
+static const char* doc_beta     = "`beta` in per-coordinate FTRL-Proximal algorithm";
+static const char* doc_lambda1  = "L1 regularization parameter";
+static const char* doc_lambda2  = "L2 regularization parameter";
+static const char* doc_d        = "Number of bins to be used for the hashing trick";
+static const char* doc_n_epochs = "Number of epochs to train a model";
+static const char* doc_inter    = "Parameter that controls if feature interactions to be used or not";
+
```

Comment:
My idea was to iterate over this vector to handle parameters automatically at some point. I really don't like that we have to specify `alpha`, `beta`, and others many times in different parts of `py::Ftrl` class. May not be possible to automate it by 100%, but at would like to reduce number of duplicates at least.

Reply:
I'm not sure how this could work? Parameters have different types, different value constraints, tied to different fields in the Ftrl class, their getters/setters must each be separate functions, etc. If you do come up with an interesting approach here, it would be very quick to put the docstrings into an array.

Reply:
I'm not sure either for the moment. Some thoughts:
- most of getters and setters follow exactly the same pattern in both their names (`get_smth`  /`set_smth`), and their implementation (simple return / assignment). If there is a getter or setter that is different, it could be supplied as a custom method. Others, in my case this is a majority of parameters, could probably be automated;
- if we have a set of predefined validation rules (like sign check, equal to zero check, within an interval, etc.) this will probably cover most of the constrains. Again, if there is something that requires a custom validator, it could be set as an external method;
- yes, types should be supplied in the same array along with the parameter names and their descriptions.